fileio/simpleio.go

```diff
@@ -27,28 +50,48 @@ func (reader *SimpleReader) Read(b []byte) (n int, err error) {
 // SimpleReader will prcoess gzipped files accordinging by performing a check on the suffix
 // of the provided file.
 func NewSimpleReader(filename string) *SimpleReader {
+	var pool = sync.Pool{
+		// New creates an object when the pool has nothing available to return.
+		// New must return an interface{} to make it flexible. You have to cast
+		// your type after getting it.
+		New: func() interface{} {
+			// Pools often contain things like *bytes.Buffer, which are
+			// temporary and re-usable. In this case we have a pointer to a slice of bytes.
+			return NewLine()
+		},
+	}
 	var answer SimpleReader = SimpleReader{
-		File: MustOpen(filename),
+		file:   MustOpen(filename),
+		pool:   &pool,
+		buffer: &bytes.Buffer{},
 	}
 	switch true {
 	case strings.HasSuffix(filename, ".gz"):
-		gzipReader, err := gzip.NewReader(answer.File)
+		gzipReader, err := gzip.NewReader(answer.file)
 		common.ExitIfError(err)
 		answer.Reader = bufio.NewReader(gzipReader)
 	default:
-		answer.Reader = bufio.NewReader(answer.File)
+		answer.Reader = bufio.NewReader(answer.file)
 	}
 	return &answer
 }
 
-// ReadLine will return a slice of bytes for each line and provided this function is called within a loop,
-// will return bool to continue reading. Please be aware that the function will call close on the file once
-// the reader encounters EOF.
-func ReadLine(reader *SimpleReader) ([]byte, bool) {
-	curr, err := reader.ReadBytes('\n')
+// ReadLine will return a bytes.Buffer pointing to the internal slice of bytes. Provided this function is called within a loop,
+// the function will read one line at a time, and return bool to continue reading. Important to note the buffer return points to
+// the internal slice belonging to the reader, meaning the slice will be overridden if the data is not copied. Please be aware the
+// reader will call close on the file once the reader encounters EOF.
+func ReadLine(reader *SimpleReader) (*bytes.Buffer, bool) {
+	var err error
+	curr := reader.pool.Get().(*Line)
```

Comment:
It seems like this is used to allocate and reuse a single Line (aka byte slice) each time we call the function.  Is it simpler/faster to remove the complexity of the memory pool (where we never know how many objects will be allocated or freed) and just have the reader hold a slice of byte that is re-used?  Essentially, instead of having a pool that you give and take "curr" from.  You just have "curr" in the reader.  I think this works since you only need one.

Reply:
That is a very good point, let me test this real quick

Reply:
pool definitely help, what is unclear to me is if reusing a single Line (aka byte slice) actually helps.

Reply:
be sure to checkout simpleio_test.go. I added what I thought was an interesting benchmark. If you are benchmarking purely runtime, using the pool is faster, but only by a few nanoseconds (not noticeably faster).
`BenchmarkSimpleReader-8    	       1	1062399970 ns/op`
`BenchmarkEasyReader-8      	       1	 1133461460 ns/op`
But if you look like the memory allocations on each line, pool wins by a landslide:
`BenchmarkSimpleReader-8    	       1	1062399970 ns/op	  291272 B/op	    4561 allocs/op`
`BenchmarkEasyReader-8      	       1	1133461460 ns/op	217882672 B/op	 3404566 allocs/op`

Reply:
I am not sure why the sync.pool would be faster on your machine.  Maybe something about the scheduling on processors is different on a mac?  This might be beyond my understanding of modern hardware.  It seems like it would be more time to keep putting the byte slice into the sync.pool and then asking for it back again.  I did an almost exact copy-and-paste of your simple reader, but without the sync.pool and just have a byte slice instead.  Because gz is so slow, everything benchmarks pretty similar on gzipped files, but I see a difference on my linux machine for uncompressed files.  Can you take a quick look at the benchmarks at the bottom of simpleio_test.go on branch: craigTestReader?

Reply:
oh there is actually a huge difference. With your benchmark I actually see a big difference for gzip files as well. I'm trying to figure out how I had something different yesterday, the only difference I can think of is that i was using a pointer to a slice byte

Reply:
the small modification you suggested actually cuts gzip reading in half as well. What I was trying to accomplish with the pool was I wanted to read the data, analyze it and/or write the results to file; put the buffer back before making a copy. I was trying to figure out a way to pass back a pointer to the underlying byte slice without actually copying what I was readying.